sklearn/manifold/tests/test_t_sne.py

```diff
@@ -825,8 +825,42 @@ def test_bh_match_exact():
         n_iter[method] = tsne.n_iter_
 
     assert n_iter['exact'] == n_iter['barnes_hut']
-    assert_array_almost_equal(X_embeddeds['exact'], X_embeddeds['barnes_hut'],
-                              decimal=3)
+    assert_allclose(X_embeddeds['exact'], X_embeddeds['barnes_hut'], rtol=1e-4)
+
+
+def test_parallel_gradient_bh_match_iterative():
+    # check that the bh gradient with different num_threads gives the same
+    # results
+
+    n_features = 10
+    n_samples = 30
+    n_components = 2
+    degrees_of_freedom = 1
+
+    kl_and_grad = {}
+
+    angle = 3
+    perplexity = 5
+
+    random_state = check_random_state(0)
+    data = random_state.randn(n_samples, n_features).astype(np.float32)
+    params = random_state.randn(n_samples, n_components)
+
+    n_neighbors = n_samples - 1
+    distances_csr = NearestNeighbors().fit(data).kneighbors_graph(
+        n_neighbors=n_neighbors, mode='distance')
+    P_bh = _joint_probabilities_nn(distances_csr, perplexity, verbose=0)
+    for num_threads in [1, 2, 4]:
+        kl_and_grad[num_threads] = _kl_divergence_bh(
+            params, P_bh, degrees_of_freedom, n_samples, n_components,
+            angle=angle, skip_num_points=0, verbose=0, num_threads=num_threads)
+
+        if num_threads != 1:
+
+            assert_allclose(kl_and_grad[1][0], kl_and_grad[num_threads][0],
+                            rtol=1e-5)
+            assert_allclose(kl_and_grad[1][1], kl_and_grad[num_threads][1],
+                            rtol=1e-5)
```

Comment:
For maximum readability I'd rewrite it that way:
```python
kl_sequential, grad_sequential = _kl_divergence_bh(
    params, P_bh, degrees_of_freedom, n_samples, n_components,
    angle=angle, skip_num_points=0, verbose=0, num_threads=1)
for num_threads in [2, 4]:
    kl_multithread, grad_multithread = _kl_divergence_bh(
        params, P_bh, degrees_of_freedom, n_samples, n_components,
        angle=angle, skip_num_points=0, verbose=0, num_threads=num_threads)

    assert_allclose(kl_multithread, kl_sequential, rtol=1e-5)
    assert_allclose(grad_multithread, grad_sequential, rtol=1e-5)
```

Reply:
Although I think 1e-5 is fine, the tests fail with higher precision ?

Reply:
I reduced to `1e-6` for the kl divergence value (`1e-7` fails) and removed the parameter for the gradient as it does not fail without it.